Migrate nonlocal games 1596 task1 - quantum

ggridin opened the pull request:
The link to the issue: https://github.com/microsoft/qsharp/issues/1596

This pull request is covering
1. CHSH quantum tasks (2.2, 2.4; 2.5 is converted to a demo)
2. CHSH discussion
3. Few changes in CHSH classical

billti commented on katas/content/nonlocal_games/chsh_quantum_bob_strategy/solution.md:
```diff
@@ -0,0 +1,11 @@
+Measuring a qubit in the $\theta$ basis is the same as rotating the qubit by $\theta$, clockwise, and then making a standard measurement in the Z basis.
+
+To implement the described transformation in Q#, we need to rotate the qubit by $\frac{\pi}{8}$ clockwise if `bit = false` or counterclockwise if `bit = true` and then perform a measurement.
```
Minor nit, but if it looks like a code condition, then it should probably use comparison operators (`bit == false`), rather than assignment (`bit = false`).

ggridin replied:
Updated "=" -> "==" in "Cosmetic updates", together with index.md EOF newline.

swernli commented:
> > chsh_quantum_bob_strategy (task 2.2) is added. please note that base profile has error "cannot use dynamic bool value".- is there fix or workaround for this problem?
> 
> * I think we would be fine with that, since katas use `unrestricted` profile.
> 
> > Quantinuum emulator profile has error "cannot use dynamic bool/string value".
> 
> * Similarly here, katas use `unrestricted` profile. See this example: https://quantum.microsoft.com/en-us/experience/quantum-katas?kataId=deutsch_algo&sectionId=deutsch_algo__e2e
>   @billti , @swernli, @tcNickolas  do you think there are plans to use `Base` profile for katas?

Yes, that's expected. The katas exercises run Unrestricted and do not need to be compatible with QIR Base profile.

tcNickolas commented on katas/content/nonlocal_games/chsh_quantum_bob_strategy/index.md:
```diff
@@ -0,0 +1,8 @@
+- Bob's half of Bell pair he shares with Alice.
+
+**Goal:**
+Measure Bob's qubit in the $\frac{\pi}{8}$ basis if his bit is 0 (false), or the $-\frac{\pi}{8}$ basis if his bit is 1 (true), and return the result.
```
We haven't defined the angular basis in this kata yet, so I'm not sure the learner will be able to follow what we're asking them to do. We should include the definition of the basis here, probably by spelling out the basis states that form it explicitly and their mapping to the measurement outcomes (we do something similar in the Measurements in Single-Qubit Systems kata, task "Distinguishing Orthogonal States: 1")

ggridin replied:
Hint is added and Solution is updated (pattern is borrowed from "Distinguishing Orthogonal States: 1" )

tcNickolas commented on katas/content/nonlocal_games/chsh_quantum_bob_strategy/solution.md:
```diff
@@ -0,0 +1,11 @@
+Measuring a qubit in the $\theta$ basis is the same as rotating the qubit by $\theta$, clockwise, and then making a standard measurement in the Z basis.
+
+To implement the described transformation in Q#, we need to rotate the qubit by $\frac{\pi}{8}$ clockwise if `bit == false` or counterclockwise if `bit == true` and then perform a measurement.
+We can do the rotation using the previous task (note the negation of the boolean parameter we need to do).
```
Since we merged the two tasks, we can't refer to the previous task any longer
```suggestion
We can do the rotation using the Ry gate (note the negation of the Boolean parameter we need to do).
```

ggridin replied:
Updated. Macro R_y is used instead of "Ry gate"

tcNickolas commented on katas/content/nonlocal_games/examples/CHSHGameDemo.qs:
```diff
@@ -0,0 +1,99 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+
+//////////////////////////////////////////////////////////////////////
+// This file contains reference solutions to all above exercises.
+// We recommend that you try to solve the exercises yourself first,
+// but feel free to look up the solution if you get stuck.
+//////////////////////////////////////////////////////////////////////
+
```
We don't include the license headers in the katas demos, and we shouldn't keep the comment from the classic katas, it doesn't really make sense in the modern katas context
```suggestion
```

ggridin replied:
All comments and license headers are removed from demo. Demo is refactored.

tcNickolas commented on katas/content/nonlocal_games/examples/CHSHGameDemo.qs:
```diff
@@ -0,0 +1,99 @@
+//////////////////////////////////////////////////////////////////////
+
+namespace Quantum.Kata.CHSHGame {
+
```
This is a very long code for a demo, and it will take a lot of scrolling for the learner to get to the interesting bits. Let's try to trim it down while keeping it interesting:

* Microsoft.Quantum.Intrinsic is open by default so we don't need to open it explicitly
* We don't need comments explaining the code from the exercises we've already covered or the /// headers
* We can merge CreateEntangledPair, PlayClassicCHSH, and PlayQuantumCHSH into the main operation - they made sense as separate operations in the classic kata when they were all separate tasks, but now that they're not and we need to copy all code from the previous tasks to the demo, having them as operations just takes up extra space, since they all are called just once
* Since the win probability is not 100%, the most interesting part of the demo is to show not the results of a single run, but to aggregate them across a lot of runs and show the difference of probabilities. Let's do a 1000 runs and compute the probabilities for classical and quantum outcomes for them instead of printing individual outcomes for each input

ggridin replied:
All comments and license headers are removed from demo. Demo is refactored.

tcNickolas reviewed (changes requested):
Apologies for the delay reviewing this PR, I had a very busy week :-(
Here are the comments for most of the changes, except for the ones in the root index.md - I'll review that file a bit later today.

Thank you!

tcNickolas commented on katas/content/nonlocal_games/index.md:
```diff
@@ -53,9 +51,147 @@ To start with, let's take a look at how you would play the classical variant of
+
+- When Bob's bit is 0, he rotates his qubit clockwise by $\pi/8$, which corresponds to the operator
+
+$$\begin{bmatrix}
+    0.9239 & 0.3827 & 0 & 0\\
+    -0.3827 & 0.9239 & 0 & 0\\
+    0 & 0 & 0.9239 & 0.3827\\
+    0 & 0 & -0.3827 & 0.9239
+\end{bmatrix}$$
+This performs the $R_y$ rotation by $\pi/8$ radians clockwise on Bob's qubit while leaving Alice's qubit unchanged.
```
I think these two lines need to be indented so that they're treated as part of the first bullet point rather than separate paragraphs, and the matrix in the second bullet point as well

ggridin replied:
Indentation is added.
Screen shot is attached.
![index_indentation](https://github.com/user-attachments/assets/839e2009-81a8-4804-b6eb-4eb5ab1d58fd)

tcNickolas commented on katas/content/nonlocal_games/index.md:
```diff
@@ -53,9 +51,147 @@ To start with, let's take a look at how you would play the classical variant of
+
+## Case 2: Alice holds bit 1
+
+When Alice holds bit 1, she measures in basis X (or, equivalently, Hadamard-transforms her qubit, leaving Bob's be, before making her Z-basis measurement).  This corresponds to applying the operator
```
```suggestion
When Alice holds bit 1, she measures in basis X (or, equivalently, Hadamard-transforms her qubit, leaving Bob's qubit unchanged, before making her Z-basis measurement).  This corresponds to applying the operator
```

tcNickolas commented on katas/content/nonlocal_games/index.md:
```diff
@@ -53,9 +51,147 @@ To start with, let's take a look at how you would play the classical variant of
+    "title": "Discussion: Probability of Victory for Quantum Strategy"
+})
+
+The above quantum strategy adopted by Alice and Bob offers a win rate of $\frac{2 + \sqrt{2}}{4}$, or about 85.36%. Let's see why this is the case.
```
The formatting will look a bit nicer if we render the probabilities as formulas; this will make them stand out from the text, same as the numbers that lead to them do later in the discussion

ggridin replied:
This and other probabilities (in current kata) were change to formula-based.
I'll follow this approach for other games, where probability have to be stand out.

tcNickolas commented on katas/content/nonlocal_games/index.md:
```diff
@@ -53,9 +51,147 @@ To start with, let's take a look at how you would play the classical variant of
+
+> As a reminder, probability "wavefunction" for a two-qubit state is given by the following length-4 vector of amplitudes:
+> $$\begin{bmatrix}\psi_{00}\\\psi_{01}\\\psi_{10}\\\psi_{11}\end{bmatrix}$$
+> $|\psi_{ij}|^2$ gives the probability of observing the corresponding basis state $|ij\rangle$ upon measuring the qubit pair.
```
Please use \ket here too instead of \rangle

tcNickolas commented on katas/content/nonlocal_games/index.md:
```diff
@@ -53,9 +51,147 @@ To start with, let's take a look at how you would play the classical variant of
+## Side notes
+
+- If Bob never rotated his qubit, their entangled state would remain the Bell state if Alice held bit 0 and the state corresponding to $\frac12 \big(|00\rangle + |01\rangle + |10\rangle - |11\rangle\big)$ if Alice held bit 1.
+
+While she and Bob would have a 100% success probability against Alice's 0 bit, they would have only a 50% chance of success if she held bit 1, and thus their win chance would revert to the 75% of the classical strategy again.
```
This needs to be part of the first bullet point; we need to either remove the empty line or indent the line after it

ggridin replied:
The sentence is merge with first bullet point (empty line is removed).
Screenshot is attached.
![index_paragraph_merge](https://github.com/user-attachments/assets/9ba3e7f9-aae0-44c2-a5f4-ce4a155e5da4)

tcNickolas commented on katas/content/nonlocal_games/index.md:
```diff
@@ -53,9 +51,147 @@ To start with, let's take a look at how you would play the classical variant of
+- Allocate two qubits and prepare a Bell state on them.
+- Send one of the qubits to Alice and another to Bob (this step is "virtual", not directly reflected in Q# code, other than making sure that Alice and Bob each act on their qubit only).
+- Have Alice and Bob perform their measurements on their respective qubits using `AliceQuantum` and `BobQuantum` operations.
+- Release(reset) used qubits
```
```suggestion
- Reset used qubits to $\ket{0}$ before they are released.
```

tcNickolas commented on katas/content/nonlocal_games/index.md:
```diff
@@ -53,9 +51,147 @@ To start with, let's take a look at how you would play the classical variant of
+- Release(reset) used qubits
+- Return their measurement results.
+
+In the example below you can compare classical and quantum results: first boolean value indicates win for classical Alice and Bob, second boolean is win for quantum Alice and Bob.
```
This will need to be updated to match the changes to the demo code

ggridin replied:
Updated to 
"In the example below you can compare winning percentage of classical and quantum games."

tcNickolas commented on katas/content/nonlocal_games/index.md:
```diff
@@ -53,9 +51,147 @@ To start with, let's take a look at how you would play the classical variant of
+- Return their measurement results.
+
+In the example below you can compare classical and quantum results: first boolean value indicates win for classical Alice and Bob, second boolean is win for quantum Alice and Bob.
+You may play with the code and check if there is a difference in results when
+1. referee picks non-random bits
+2. Bob's qubit is measured first
+3. Alice and Bob get not entangled qubit pair
```
Nice examples!! They'll need to be a separate paragraph for better text flow
```suggestion
You may play with the code and check if there is a difference in results when
1. The referee picks non-random bits. How can the referee minimize Alice and Bob's win probability?
2. Bob's qubit is measured first.
3. Alice and Bob get unentangled qubit pair.
```

ggridin replied:
Suggestions were indented ('>' is used).
Screenshot of result is attached.
![index_suggestion_indentation](https://github.com/user-attachments/assets/0a901b00-1d4b-4335-b609-aedd2e00f8e0)

tcNickolas reviewed (changes requested):
Finished reading through the index.md, posting the rest of the comments.

Additionally, could you please edit the PR description to move out everything starting with "other PRs will follow"? These details are good to include for the reviewer while the PR is in progress, but our merge system is set up in a way that includes the whole PR description in the commit message, and they're not necessary there, just making the commit info look bulky. For future PRs, these details can be made into the first comment on the PR rather than a part of the description.

Thank you!!

ggridin commented on katas/content/nonlocal_games/chsh_quantum_alice_strategy/Verification.qs:
```diff
@@ -0,0 +1,46 @@
+            // the correct answer, reduces probability of false positives
+            if (Kata.AliceQuantum(false, q) != false) {
+                Message("|0⟩ not measured as false");
+                Reset(q);
```
@tcNickolas
Reset(q) is called at the bottom of the for loop at the line 41.
Honestly, I do not like multiple "Reset(q);" statements - it easy to miss/forget one.
However, I cannot find the good code pattern that reduces the number of qubit resets.

tcNickolas replied:
Yes, but each loop body calls the kata solution four times, reusing the qubit in whatever state the previous solution call left it to prepare the next input state. Without resetting it, you don't know what input you're giving to later solution calls

ggridin replied:
Thank you for good catch - I missed the case.
Verification.qs is refactored (slightly) to fix the problem. One "Reset(q)" was dropped!

ggridin commented:
Additional details of this PR:
1. Section "CHSH Classical" is merged with "CHSH Game"
   - DrawRandomBool is used instead of DrawRandomInt
2. chsh_quantum_alice_strategy (task 2.2) is added
3. chsh_quantum_bob_strategy (task 2.2) is added
   - please note that base profile has error "cannot use dynamic bool value".- is there fix or workaround for this problem?
4. Discussion content is copied from previous kata version
5. Task 2.5 is significantly refactored - needs thoughtful review.
   The benefits of the change:
   - Demo is comparison of classic and quantum results - which is interesting to observe.
   - Entangled qubits are created before the game.
6. Few cosmetic changes in nonlocal_games/index.md suggested by vscode markdown linter

Testing done:
1. Build "python ./build.py --npm" is successful
2. Placeholder cases are syntactically correct but failing verifications.
   Solutions are successful.
3. Some other failure scenarios were tested manually.
4. Content screenshots are attached.
5. Demo is tested in Azure Quantum coding (in-memory simulator): https://quantum.microsoft.com/en-us/experience/quantum-coding
   - Quantinuum emulator profile has error "cannot use dynamic bool/string value".

![CHSH Alice Quantum Strategy](https://github.com/user-attachments/assets/fa501b1a-bc79-4fd7-b217-86633f8ec5fd)
![CHSH Base profile error](https://github.com/user-attachments/assets/d67737dd-a1d8-489d-9882-7b614c725448)
![CHSH Bob Quantum Strategy](https://github.com/user-attachments/assets/01a18c80-133b-4a19-b644-72914d5d9c28)
![CHSH Classic Lesson title is removed](https://github.com/user-attachments/assets/dd0470ff-9874-40f3-b087-693e059d2897)
![CHSH Discussion Case 1](https://github.com/user-attachments/assets/6ccab991-9ee5-4e87-8acb-df51805a9687)
![CHSH Discussion Case 2](https://github.com/user-attachments/assets/140e930d-9622-45c7-aa9c-32487d3fbe49)
![CHSH Running CHSH game - results](https://github.com/user-attachments/assets/1bdc89ef-756b-4569-8753-4f00c803bdd8)
![CHSH Running CHSH game - text](https://github.com/user-attachments/assets/93550226-6b5f-4ffc-afaa-8b80935f0cc9)

ggridin commented on katas/content/nonlocal_games/chsh_quantum_alice_strategy/Solution.qs:
```diff
@@ -0,0 +1,15 @@
+    operation AliceQuantum (bit : Bool, qubit : Qubit) : Bool {
+
+        // Measure in sign basis if bit is 1, and
+        // measure in computational basis if bit is 0
```
Fixed.

ggridin commented on katas/content/nonlocal_games/chsh_quantum_alice_strategy/Solution.qs:
```diff
@@ -0,0 +1,15 @@
+        }
+        else {
+            let q = MResetZ(qubit);
+            return (q == One);
```
Fixed.

ggridin commented on katas/content/nonlocal_games/chsh_quantum_alice_strategy/index.md:
```diff
@@ -0,0 +1,11 @@
+
+**Input:**
+
+- Alice's starting bit (X)
```
Fixed.

ggridin commented on katas/content/nonlocal_games/chsh_quantum_alice_strategy/solution.md:
```diff
@@ -0,0 +1,12 @@
+In Q#, you can perform measurements in a specific basis using either the
+[Measure operation](https://docs.microsoft.com/qsharp/api/qsharp/microsoft.quantum.intrinsic.measure)
+or convenient shorthands for measure-and-reset-to-$|0\rangle$ sequence of operations
```
Fixed. Updated all remaining \rangle macros to \ket

ggridin commented on katas/content/nonlocal_games/chsh_quantum_bob_strategy/Solution.qs:
```diff
@@ -0,0 +1,10 @@
+    open Microsoft.Quantum.Math;
+
+    operation BobQuantum (bit : Bool, qubit : Qubit) : Bool {
+
```
Fixed.

ggridin commented on katas/content/nonlocal_games/chsh_quantum_alice_strategy/index.md:
```diff
@@ -0,0 +1,11 @@
+- Alice's half of Bell pair she shares with Bob.
+
+**Goal:**
+  Measure Alice's qubit in the Z basis if her bit is 0 (false), or the X basis if her bit is 1 (true), and return the result.
```
Fixed.

ggridin commented on katas/content/nonlocal_games/chsh_quantum_bob_strategy/index.md:
```diff
@@ -0,0 +1,8 @@
+**Input:**
+
+- Bob's starting bit (Y)
```
Fixed.

ggridin commented on katas/content/nonlocal_games/chsh_quantum_alice_strategy/Verification.qs:
```diff
@@ -0,0 +1,46 @@
+            // repeat 4 times since we are testing a measurement and wrong basis still might get
+            // the correct answer, reduces probability of false positives
+            if (Kata.AliceQuantum(false, q) != false) {
+                Message("|0⟩ not measured as false");
```
Fixed this and 3 remaining messages.

ggridin commented:
> Finished reading through the index.md, posting the rest of the comments.
> 
> Additionally, could you please edit the PR description to move out everything starting with "other PRs will follow"? These details are good to include for the reviewer while the PR is in progress, but our merge system is set up in a way that includes the whole PR description in the commit message, and they're not necessary there, just making the commit info look bulky. For future PRs, these details can be made into the first comment on the PR rather than a part of the description.
> 
> Thank you!!

@tcNickolas 
Thank you for the detailed and thoughtful review!
I believe that I fixed all mentioned issues. I hope I wont miss something, considering amount of changes.

In the current PR, I've updated PR description according your suggestion and moved the details to one of the comments.
I'll do the same shortly for https://github.com/microsoft/qsharp/pull/1783 too

ggridin commented on katas/content/nonlocal_games/chsh_quantum_bob_strategy/index.md:
```diff
@@ -0,0 +1,8 @@
+- Bob's half of Bell pair he shares with Alice.
+
+**Goal:**
+Measure Bob's qubit in the $\frac{\pi}{8}$ basis if his bit is 0 (false), or the $-\frac{\pi}{8}$ basis if his bit is 1 (true), and return the result.
```
Updated.

tcNickolas commented on katas/content/nonlocal_games/chsh_quantum_alice_strategy/Solution.qs:
```diff
@@ -0,0 +1,14 @@
+namespace Kata {
+    operation AliceQuantum (bit : Bool, qubit : Qubit) : Bool {
+        // Measure in sign basis if bit is 1, and
+        // measure in computational basis if bit is 0
```
We don't need comments in the solutions
```suggestion
```

tcNickolas commented on katas/content/nonlocal_games/chsh_quantum_alice_strategy/solution.md:
```diff
@@ -0,0 +1,12 @@
+In Q#, you can perform measurements in a specific basis using either the
+[Measure operation](https://docs.microsoft.com/qsharp/api/qsharp/microsoft.quantum.intrinsic.measure)
+or convenient shorthands for measure-and-reset-to-$\ket{0}$ sequence of operations
+[MResetZ](https://docs.microsoft.com/qsharp/api/qsharp/microsoft.quantum.measurement.mresetz) and
+[MResetX](https://docs.microsoft.com/qsharp/api/qsharp/microsoft.quantum.measurement.mresetx).
```
Looks like these are links to old documentation, not sure why I didn't notice that on the first read, sorry
```suggestion
In Q#, you can perform measurements in a specific basis using either the
[Measure operation](https://learn.microsoft.com/qsharp/api/qsharp-lang/microsoft.quantum.intrinsic/measure)
or convenient shorthands for measure-and-reset-to-$\ket{0}$ sequence of operations
[MResetZ](https://learn.microsoft.com/qsharp/api/qsharp-lang/microsoft.quantum.measurement/mresetz) and
[MResetX](https://learn.microsoft.com/qsharp/api/qsharp-lang/microsoft.quantum.measurement/mresetx).
```

tcNickolas commented on katas/content/nonlocal_games/chsh_quantum_bob_strategy/index.md:
```diff
@@ -0,0 +1,14 @@
+if his bit is 1 (true) and return the measurement result as a Boolean value: map `Zero` to false and `One` to true.
+The state of the qubit after the operation does not matter.
+
+<details>
+<summary><strong>Need a hint?</strong></summary>
+Measuring a qubit in the $\theta$ basis is the same as rotating the qubit by $\theta$, clockwise, and then making a standard measurement in the Z basis.
+</details>
```
I think this should be part of the problem definition rather than a hint
```suggestion
Measuring a qubit in the $\theta$ basis is the same as rotating the qubit by $\theta$, clockwise, and then making a standard measurement in the Z basis.
```

tcNickolas commented on katas/content/nonlocal_games/index.md:
```diff
@@ -53,9 +51,147 @@ To start with, let's take a look at how you would play the classical variant of
+    0 & 0 & 0.9239 & 0.3827\\
+    0 & 0 & -0.3827 & 0.9239
+\end{bmatrix}$$
+>This performs the $R_y$ rotation by $\pi/8$ radians clockwise on Bob's qubit while leaving Alice's qubit unchanged.
```
This looks odd as a note. Oddly, if I try to indent the formula above, it gives me a KaTeX parser error, so I guess that won't work, but let's leave it as plain text.
```suggestion
This performs the $R_y$ rotation by $\pi/8$ radians clockwise on Bob's qubit while leaving Alice's qubit unchanged.
```

tcNickolas commented on katas/content/nonlocal_games/examples/CHSHGameDemo.qs:
```diff
@@ -0,0 +1,53 @@
+
+    @EntryPoint()
+    operation CHSH_GameDemo() : Unit {
+        use aliceQubit = Qubit();
+        use bobQubit = Qubit();
```
Minor, but this can be shortened a bit if we allocate a tuple right away
```suggestion
        use (aliceQubit, bobQubit) = (Qubit(), Qubit());
```

tcNickolas reviewed (approved):
Looks great! And the updated demo is much more illustrative of the differences between classical and quantum strategies!

I left a few minor comments, but I shouldn't need to take another look. Just need a second reviewer's signoff.

Thank you!
